lib/io/src/i2c.rs

```diff
@@ -22,3 +21,40 @@ pub trait HypedI2c {
         data: u8,
     ) -> Result<(), I2cError>;
 }
+
+pub mod mock_i2c {
+    use heapless::FnvIndexMap;
+
+    /// A fixed-size map of I2C values, indexed by device address and register address
+    type I2cValues = FnvIndexMap<(u8, u8), u8, 16>;
+
+    /// A mock I2C instance which can be used for testing
+    pub struct MockI2c {
+        values: I2cValues,
+    }
+
+    impl crate::i2c::HypedI2c for MockI2c {
+        /// Reads a byte by looking up the device address and register address in the map
+        fn read_byte(&mut self, device_address: u8, register_address: u8) -> Option<u8> {
+            self.values
+                .get(&(device_address, register_address))
+                .copied()
+        }
+
+        /// Always succeeds, but does nothing
```

Comment:
something you could do here if you care is store the writes that occur and have a getter, then you could test that the correct writes occur

Reply:
and/or have a list of write results, so you can check failure cases again